feat: use mock authentication for local dev

tchaffee opened the pull request:
<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [ ] My article does not contain shortened URLs or affiliate links.

Closes #17382

tchaffee committed:
feat: use mock authentication for local dev

tchaffee commented on package.json:
```diff
@@ -9,6 +9,7 @@
     "lint": "echo 'Warning: TODO - Define Linting.'",
     "pretest-ci": "npm-run-all -s lint bootstrap",
     "seed": "cross-env DEBUG=fcc:* node ./tools/scripts/seed/seedChallenges",
+    "seed-auth-user": "cross-env DEBUG=fcc:* node ./tools/scripts/seed/seedAuthUser",
```
I could use some guidance on how to set this part up.

tchaffee replied:
@raisedadead

raisedadead replied:
Sure, I'll propose a few changes in the areas, which you can add or accept.

tchaffee commented on tools/scripts/seed/seedAuthUser.js:
```diff
@@ -0,0 +1,101 @@
+const path = require('path');
```
Not sure if this file even belongs in this dir? @raisedadead can you please look at the `package.json` file in this dir? It's pretty specific to seeding the challenges. Do you want me to change that? Or put this file in its own dir? I'm open to suggestions on how to organize this.

raisedadead replied:
Yes, you are correct. These things weren't cleaned up well. I propose we put all the seeding scripts in this same directory. Most likely the devDeps around this scripts are going to be more or less same. It makes sense to keep them together.

So, indeed the `package.json` in the directory needs cleanup. That's not a concern of this PR per se, yet I welcome to restructure/change it as you please. Thanks for taking the additional effort cleaning it up. If you get too confused on that, do not worry I can make the changes.

Essentially we would need to be able to reference this from the root.

tchaffee commented:
@raisedadead This is complete now. But I could use your help in organizing parts of this per the previous comments.

tchaffee commented:
@raisedadead I just remembered I forgot to seed the `AccessToken` collection. But when I tested it without that record, everything worked fine. Maybe we don't need that record using my solution? If there is something I'm missing and we do need it, let's discuss so I can understand why. Easy enough to add it to the seeding process.

raisedadead commented:
Hi @tchaffee, thanks for preparing this PR. I'll pull this down and get back to you.

raisedadead commented on package.json:
```diff
@@ -9,6 +9,7 @@
     "lint": "echo 'Warning: TODO - Define Linting.'",
     "pretest-ci": "npm-run-all -s lint bootstrap",
     "seed": "cross-env DEBUG=fcc:* node ./tools/scripts/seed/seedChallenges",
+    "seed-auth-user": "cross-env DEBUG=fcc:* node ./tools/scripts/seed/seedAuthUser",
```
```suggestion
    "seed:auth-user": "cross-env DEBUG=fcc:* node ./tools/scripts/seed/seedAuthUser",
```

raisedadead replied:
This is just to be able to run all the commands in tandem with [`npm-run-all`](https://www.npmjs.com/package/npm-run-all)

raisedadead commented on package.json:
```diff
@@ -9,6 +9,7 @@
     "lint": "echo 'Warning: TODO - Define Linting.'",
     "pretest-ci": "npm-run-all -s lint bootstrap",
     "seed": "cross-env DEBUG=fcc:* node ./tools/scripts/seed/seedChallenges",
```
```suggestion
    "seed": "npm-run-all -s seed:*",
    "seed:curriculum": "cross-env DEBUG=fcc:* node ./tools/scripts/seed/seedChallenges",
```

raisedadead replied:
This is just to be able to run all the commands in tandem with [`npm-run-all`](https://www.npmjs.com/package/npm-run-all)

raisedadead commented on api-server/server/boot/authentication.js:
```diff
@@ -25,15 +29,34 @@ module.exports = function enableAuthentication(app) {
+  // Use a local mock strategy for signing in if we are in dev mode.
+  // Otherwise we use auth0 login. We use a string for 'true' because values
+  // set in the env file will always be strings and never boolean.
+  if (process.env.LOCAL_AUTH === 'true') {
```
```suggestion
  if (process.env.LOCAL_MOCK_AUTH === 'true') {
```

Sorry about the nit-pick. Let's make it a little bit less generic, and avoid potential conflict in future? Not that I am aware of that there is anything in the code or underlying libraries would use `LOCAL_AUTH`, it does help to be specific, while we are at it?

tchaffee replied:
The reason I used `LOCAL_AUTH` is that you already have it in the `sample.env`. Do you want to go with `LOCAL_MOCK_AUTH` and I'll update the `sample.env` too?

raisedadead commented:
I am quite happy with the logic flow. I am yet to test this out. 

Unrelated side notes: (*This is not at all a concern of this PR and must be taken later*)

 - One this that is killing me is at the minute is the `user.insertOne`. Its messy because we are replicating the schema in the code. I am not a fan of that.
-  The ideal way would be to spin up loopback and do the seeding with its APIs. i.e `User.create`. This is so, that we do not worry about the default values in the schema. Loopback will add/update the keys for us and then we just do a `User.Update` on specific things.
- This should ideally be done for challenge collection as well.

Again just for discussion, do not worry about it right now. I am pretty sure we abandoned it for some reason that I can't recall.

raisedadead commented:
> I just remembered I forgot to seed the AccessToken collection. But when I tested it without that record, everything worked fine. Maybe we don't need that record using my solution? 

I'll get back with a update when I am done testing. Thanks for your patience.

raisedadead reviewed (changes requested):
I forgot! Could you please update the [`sample.env`](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/sample.env) with this default value for local env variable.

By default the setup starts in development mode.

raisedadead commented:
> > I just remembered I forgot to seed the AccessToken collection. But when I tested it without that record, everything worked fine. Maybe we don't need that record using my solution? 
> 
> I'll get back with a update when I am done testing. Thanks for your patience.
> 

So, what I can see from [your changes](https://github.com/freeCodeCamp/loopback-component-passport/blob/ac221374736f6a741729d7e399f7c4a260077c35/lib/passport-configurator.js#L325-L333):

```js
            // This is the same format as would be returned by Auth0.
            // The id comes from a record that should already be in the db
            // due to seeding at development time.
            const accessToken = {
              id: 'nyynsTyQrmXNo8APMKiG7SAkSvsBiC1XnZ8QhI2GuhfixAaV6hEpuJ5gFTNeEcJU',
              ttl: 77760000000,
              created: new Date().toISOString(),
              userId: u.id
            };
```

This is where you are by passing the an AccessToken creation with Loopback and hardcoding it. I think we are probably going to be fine using this hack 😉 !

Since, the real reason behind the local auth is to let collaborators to be able to work with pages that are hidden behind an authenticated user, this works perfectly for our needs.

tchaffee commented:
@raisedadead I added `LOCAL_MOCK_AUTH=true` to `sample.env` and changed the test in `authentication.js` to test against that value. Let me know if there is anything else outstanding. Otherwise I think this is ready for merge once someone else has tested it locally.

nikrb commented:
@tchaffee this works locally for me.

Welcome page "Goto to next challenge" button seems to be dead, but curriculum menu item works.
Clicking the dev user icon in the menu top right goes to the settings page. "Show me my public portfolio" button shows me a gatsby page missing error. "Sign me out of freecodecamp" button seems to work.

raisedadead commented:
Hi @tchaffee @nikrb 

We are still working on a couple things for the platform, and hence the delay on getting this merged. 

But just to let you know:

> ... Welcome page "Goto to next challenge" button seems to be dead ...

This is known and Stuart has already dropped a fix.

> "Show me my public portfolio" button shows me a gatsby page missing error...

This is a unfortunate limitation of Gatsby's Dev Server for the moment. 

The public profile page is actually available and you should see it just fine if you click the button that says `Preview custom 404 page`.

![image](https://user-images.githubusercontent.com/1884376/49363633-31141980-f708-11e8-828c-fc62df69291b.png)

Please note this will not be an issue when the client site is deployed to the production. You won't have the dev server intercepting the dynamic routes. We would need a fix to the Gatsby core to address this and I think they have something in works.

Until then you can simply click the button to get you to the right page when you are developing.

tchaffee commented:
@nikrb thank you for testing this.

@raisedadead I think I've made all your requested changes. Is there anything else I need to do on this or are we just waiting now for other stuff?

QuincyLarson commented:
@raisedadead Could you take another look at this and see whether it's ready to go (aside from the merge conflicts)? This could be a huge improvement to onboarding new contributors.

raisedadead committed:
fix: update custom loopback-component-passport

raisedadead committed:
fix: remove extraneous key

raisedadead reviewed (commented):
Needs updated  `@freeCodeCamp/loopback-component-passport`

raisedadead reviewed (approved):
Nice work @tchaffee !

I have taken the liberty and did some chores. Other than that this was an awesome PR to QA. Thanks for your patience with this and the hard work.

Happy contributing!
